File: go/cases_test.go
package cjkfixtures

import "testing"

// Self-verification for the exported cases. Every case states a correct result
// and the wrong result a broken handler produces; this proves both halves
// against this package's own reference helpers, so the data shipped to adopters
// is real. If a literal in cases.go is wrong, this fails.

func TestWidthCases(t *testing.T) {
	for _, c := range WidthCases {
		if got := DisplayWidth(c.Input); got != c.CorrectWidth {
			t.Errorf("%s %q: DisplayWidth = %d, want %d", c.Slug, c.Input, got, c.CorrectWidth)
		}
		if got := NaiveWidth(c.Input); got != c.NaiveWidth {
			t.Errorf("%s %q: NaiveWidth = %d, want %d", c.Slug, c.Input, got, c.NaiveWidth)
		}
	}
}

func TestEqualityCases(t *testing.T) {
	for _, c := range EqualityCases {
		if got := NaiveEquals(c.A, c.B); got != c.RawEqual {
			t.Errorf("%s: NaiveEquals(%q,%q) = %v, want %v", c.Slug, c.A, c.B, got, c.RawEqual)
		}
		// CanonicalEquals settles the canonical (NFC/NFD) layer for every row.
		if got := CanonicalEquals(c.A, c.B); got != c.NFCEqual {
			t.Errorf("%s: CanonicalEquals(%q,%q) = %v, want %v", c.Slug, c.A, c.B, got, c.NFCEqual)
		}
		// CompatFold here is a width-only fold, not full NFKC: it folds the
		// width variants (#9) but does not subsume canonical equivalence, so it
		// is only asserted on the #9 rows. A real NFKC normalizer would also
		// make the #7 rows equal, which is what NFKCEqual records.
		if c.Mode == 9 {
			if got := CompatEquals(c.A, c.B); got != c.NFKCEqual {
				t.Errorf("%s: CompatEquals(%q,%q) = %v, want %v", c.Slug, c.A, c.B, got, c.NFKCEqual)
			}
		}
		if c.NFCEqual && !c.NFKCEqual {
			t.Errorf("%s: NFC-equal implies NFKC-equal", c.Slug)
		}
	}
}

func TestDirectionCases(t *testing.T) {
	for _, c := range DirectionCases {
		if got := BaseDirection(c.Input); got != c.CorrectDir {
			t.Errorf("%s %q: BaseDirection = %q, want %q", c.Slug, c.Input, got, c.CorrectDir)
		}
		if got := NaiveDirection(c.Input); got != c.NaiveFirstCharDir {
			t.Errorf("%s %q: NaiveDirection = %q, want %q", c.Slug, c.Input, got, c.NaiveFirstCharDir)
		}
	}
}

func TestWhitespaceCases(t *testing.T) {
	for _, c := range WhitespaceCases {
		if got := UnicodeTrim(c.Input); got != c.CorrectTrimmed {
			t.Errorf("%s %q: UnicodeTrim = %q, want %q", c.Slug, c.Input, got, c.CorrectTrimmed)
		}
		if got := IsBlankUnicode(c.Input); got != c.CorrectBlank {
			t.Errorf("%s %q: IsBlankUnicode = %v, want %v", c.Slug, c.Input, got, c.CorrectBlank)
		}
		if got := AsciiTrim(c.Input); got != c.ASCIITrimmed {
			t.Errorf("%s %q: AsciiTrim = %q, want %q", c.Slug, c.Input, got, c.ASCIITrimmed)
		}
		if got := IsBlankAscii(c.Input); got != c.ASCIIBlank {
			t.Errorf("%s %q: IsBlankAscii = %v, want %v", c.Slug, c.Input, got, c.ASCIIBlank)
		}
	}
}

func TestSliceCases(t *testing.T) {
	for _, c := range SliceCases {
		switch c.Unit {
		case "byte":
			if got := ByteSlice(c.Input, c.Limit); got != c.Torn {
				t.Errorf("%s %q: ByteSlice(%d) = %q, want %q", c.Slug, c.Input, c.Limit, got, c.Torn)
			}
			if IsCleanUTF8(c.Torn) {
				t.Errorf("%s: torn %q should not be clean UTF-8", c.Slug, c.Torn)
			}
			if !IsCleanUTF8(c.Whole) {
				t.Errorf("%s: whole %q should be clean UTF-8", c.Slug, c.Whole)
			}
		case "grapheme":
			if got := CodePointSlice(c.Input, c.Limit); got != c.Torn {
				t.Errorf("%s %q: CodePointSlice(%d) = %q, want %q", c.Slug, c.Input, c.Limit, got, c.Torn)
			}
			if got := GraphemeSlice(c.Input, c.Limit); got != c.Whole {
				t.Errorf("%s %q: GraphemeSlice(%d) = %q, want %q", c.Slug, c.Input, c.Limit, got, c.Whole)
			}
			if c.Torn == c.Whole {
				t.Errorf("%s: torn and whole should differ", c.Slug)
			}
		default:
			t.Errorf("%s: unknown unit %q", c.Slug, c.Unit)
		}
	}
}

func TestEditorCases(t *testing.T) {
	for _, c := range EditorCases {
		good := ApplyEvents(NewEditor(), c.Events)
		if good.Value != c.Correct.Value || good.Submitted != c.Correct.Submitted {
			t.Errorf("%s: correct = {%q,%d}, want {%q,%d}",
				c.Slug, good.Value, good.Submitted, c.Correct.Value, c.Correct.Submitted)
		}
		bad := ApplyEvents(NewEditor(c.BrokenBug), c.Events)
		if bad.Value != c.Broken.Value || bad.Submitted != c.Broken.Submitted {
			t.Errorf("%s: broken = {%q,%d}, want {%q,%d}",
				c.Slug, bad.Value, bad.Submitted, c.Broken.Value, c.Broken.Submitted)
		}
		if bad.Value == c.Correct.Value && bad.Submitted == c.Correct.Submitted {
			t.Errorf("%s: bug %q did not change the outcome", c.Slug, c.BrokenBug)
		}
	}
}

func TestCaseCoverage(t *testing.T) {
	modes := map[int]bool{}
	for _, c := range WidthCases {
		modes[c.Mode] = true
	}
	for _, c := range EqualityCases {
		modes[c.Mode] = true
	}
	for _, c := range DirectionCases {
		modes[c.Mode] = true
	}
	for _, c := range WhitespaceCases {
		modes[c.Mode] = true
	}
	for _, c := range SliceCases {
		modes[c.Mode] = true
	}
	for _, c := range EditorCases {
		modes[c.Mode] = true
	}
	// Go mirrors every mode except #8 (a UTF-16-specific hazard with no native
	// analogue in Go's UTF-8 strings).
	for _, m := range []int{1, 2, 3, 4, 5, 6, 7, 9, 10, 11} {
		if !modes[m] {
			t.Errorf("mode %d missing from Go cases", m)
		}
	}
}

File: js/test/cases.test.js
// Self-verification for the exported cases.
//
// Every case states a correct result and the wrong result a broken handler
// produces. This file proves both halves against this repo's own reference
// helpers, so the data shipped to adopters is real, not hand-typed guesses. If
// a literal in cases.js is wrong, this fails.
import { describe, it, expect } from "vitest";
import {
  widthCases,
  equalityCases,
  directionCases,
  whitespaceCases,
  sliceCases,
  editorCases,
  applyEvents,
  cases,
  allCases,
} from "../src/cases.js";
import { displayWidth, naiveWidth } from "../src/width.js";
import { byteSlice, codePointSlice, hasReplacementChar } from "../src/bytes.js";
import { utf16Slice, isWellFormed } from "../src/surrogate.js";
import { graphemeSlice, codePointSlice as cpSlice } from "../src/graphemes.js";
import { naiveEquals, normalizedEquals } from "../src/normalize.js";
import { compatEquals } from "../src/compat.js";
import { baseDirection, naiveDirection } from "../src/bidi.js";
import { asciiTrim, unicodeTrim, isBlankAscii, isBlankUnicode } from "../src/whitespace.js";
import { createEditor } from "../src/editor.js";

describe("width cases", () => {
  it.each(widthCases)("$input -> $correctWidth", (c) => {
    expect(displayWidth(c.input)).toBe(c.correctWidth);
    expect(naiveWidth(c.input)).toBe(c.naiveWidth);
  });
});

describe("equality cases", () => {
  it.each(equalityCases)("$a vs $b", (c) => {
    expect(naiveEquals(c.a, c.b)).toBe(c.rawEqual);
    expect(normalizedEquals(c.a, c.b)).toBe(c.nfcEqual);
    expect(compatEquals(c.a, c.b)).toBe(c.nfkcEqual);
    // NFKC must subsume NFC: anything NFC-equal is NFKC-equal.
    if (c.nfcEqual) expect(c.nfkcEqual).toBe(true);
  });
});

describe("direction cases", () => {
  it.each(directionCases)("$input", (c) => {
    expect(baseDirection(c.input)).toBe(c.correctDir);
    expect(naiveDirection(c.input)).toBe(c.naiveFirstCharDir);
  });
});

describe("whitespace cases", () => {
  it.each(whitespaceCases)("$input", (c) => {
    expect(unicodeTrim(c.input)).toBe(c.correctTrimmed);
    expect(isBlankUnicode(c.input)).toBe(c.correctBlank);
    expect(asciiTrim(c.input)).toBe(c.asciiTrimmed);
    expect(isBlankAscii(c.input)).toBe(c.asciiBlank);
  });
});

describe("slice cases", () => {
  it.each(sliceCases)("$slug $input", (c) => {
    if (c.unit === "byte") {
      expect(byteSlice(c.input, c.limit)).toBe(c.torn);
      expect(hasReplacementChar(c.torn)).toBe(true);
      expect(hasReplacementChar(c.whole)).toBe(false);
    } else if (c.unit === "utf16") {
      expect(utf16Slice(c.input, c.limit)).toBe(c.torn);
      expect(isWellFormed(c.torn)).toBe(false);
      expect(isWellFormed(c.whole)).toBe(true);
    } else {
      // grapheme: the code-point slice tears the cluster, the grapheme slice keeps it.
      expect(cpSlice(c.input, c.limit)).toBe(c.torn);
      expect(graphemeSlice(c.input, c.limit)).toBe(c.whole);
      expect(c.torn).not.toBe(c.whole);
    }
  });
});

describe("editor cases", () => {
  it.each(editorCases)("$slug", (c) => {
    const good = applyEvents(createEditor(), c.events);
    expect(good.value).toBe(c.correct.value);
    expect(good.submitted).toBe(c.correct.submitted);

    const bad = applyEvents(createEditor({ bugs: [c.brokenBug] }), c.events);
    expect(bad.value).toBe(c.broken.value);
    expect(bad.submitted).toBe(c.broken.submitted);

    // The bug must actually change the outcome.
    expect([c.broken.value, c.broken.submitted]).not.toEqual([
      c.correct.value,
      c.correct.submitted,
    ]);
  });
});

describe("coverage", () => {
  it("every fixture mode 1..11 has at least one case", () => {
    const modes = new Set(allCases.map((c) => c.mode));
    for (const m of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]) {
      expect(modes.has(m), `mode ${m} missing`).toBe(true);
    }
  });

  it("the grouped object and the flat list hold the same cases", () => {
    const grouped =
      cases.width.length +
      cases.equality.length +
      cases.direction.length +
      cases.whitespace.length +
      cases.slice.length +
      cases.editor.length;
    expect(grouped).toBe(allCases.length);
  });
});
